### ISSUE TITLE
RelaxedBernoulli samples not differentiable

### ISSUE THREAD

**Issue by changhoonhahn** [DifferentiableUniverseInitiative/DHOD#1]
samples from `tensorflow_probability.distributions.RelaxedBernoulli` is not differentiable even when temperature is increased to 10^6. The gradients from the snippets below spits out nans. 

```
Mhalo = tf.convert_to_tensor(np.random.uniform(10., 15., 1000), dtype=tf.float32)
siglogm = tf.convert_to_tensor(0.2, dtype=tf.float32)
temperature = 100

def Ncen(Mmin): 
    # mean occupation of centrals
    return 0.5 * (1+tf.math.erf((Mhalo - Mmin)/siglogm))

for temp in np.linspace(0.1, 1e6, 10).astype(np.float32): 
    def _hod(Mmin): 
        bern = tfp.distributions.RelaxedBernoulli(temp, probs=Ncen(Mmin))
        return bern.sample(seed=0)

    loss = lambda mm: _hod(mm)
    val, grad = tfp.math.value_and_gradient(loss, [_Mmin])
    print(grad)
```

**Comment by changhoonhahn** [comment]
When I tried replacing `_hod` with `tfp.distributions.Normal` I didn't run into any issues...

**Comment by EiffL** [comment]
Ok, sorry it took me a while to get around to this, but I confirm you are not crazy, I can't get non-zero gradients with the relaxed Bernoulli. It is very likely an issue with some underflowing. I'll try to dig out my own implementation of the Gumbel-Softmax which was working perfectly fine. Maybe that'll point us  to what is  wrong here

**Comment by EiffL** [comment]
hehehehehhehe:
![image](https://user-images.githubusercontent.com/861591/69002060-a099ff00-08ae-11ea-972b-5fb2ce1c0cd0.png)

So yeah the problem comes from underflows when the probability for Ncen drops too close to 0. I would say one simple workaround is to clip the  probability to some `p_min`, once my brain has rebooted a little bit more might think of a cleverer solution.

I've added another demo notebook: https://github.com/DifferentiableUniverseInitiative/DHOD/blob/master/nb/dhod_centrals_gs.ipynb
This works as long as we don't sample too far away from Mmin, need to implement that clipping.

**Comment by EiffL** [comment]
Sweet, closing this issue then

**Comment by EiffL** [comment]
Before we forgetr.... @all-contributors, let's add @changhoonhahn for code
